torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -0,0 +1,191 @@
+                min_vals = [x[0] for x in constraints]
+                max_vals = [x[1] for x in constraints]
+                min_val = sorted(min_vals, reverse=True)[0]
+                max_val = sorted(max_vals, reverse=False)[0]
+
+                assert min_val <= max_val
+
+                constraints_id_to_constraint[t_id].append(
+                    ConstraintSpec(constraint_dim=dim, min_val=min_val, max_val=max_val)
+                )
+
+        return constraints_id_to_constraint
+
+    def call(self, graph_module: torch.fx.GraphModule) -> None:
+        # resets the counter
+        self.input_tracker = 0
+        self.current_gm = graph_module
+        self.constraints = self._process_constraints(get_export_meta(self.current_gm).input_shape_constraints)
+        self.example_inputs = pytree.tree_flatten(get_export_meta(self.current_gm).example_inputs)[0]
+        return super().call(graph_module)
+
+    def _insert_specialized_shapes_assert(self, arg, dims, current_inp):
+        # we don't want to get shapes from meta as they will be symbolic
+        shapes = current_inp.shape
+        for dim in dims:
+            assert_msg = (
+                f"Input #{self.input_tracker}'s dimension #{dim} size is "
+                f"specialized at {shapes[dim]}"
+            )
+            dim_node = self.call_operator(
+                torch.ops.aten.sym_size,
+                (arg, dim),
+                {},
+                NodeMetadata({}),
+            )
+            eq = self.call_operator(
+                operator.eq,
+                (dim_node, shapes[dim]),
+                {},
+                NodeMetadata({}),
+            )
+            tensor_eq = self.call_operator(
+                torch.ops.aten.scalar_tensor.default,
+                (eq,),
+                {},
+                NodeMetadata({}),
+            )
+            self.call_operator(
+                torch.ops.aten._assert_async.msg,
+                (tensor_eq, assert_msg),
+                {},
+                NodeMetadata({}),
+            )
+
+    def placeholder(self, name: str, arg, meta) -> ProxyValue:
+        arg = super().placeholder(name, arg, meta)
+        assert self.current_gm is not None
+        current_inp = self.example_inputs[self.input_tracker]
+        current_fake_mode = meta["val"].fake_mode
+        all_dims = set(range(current_inp.dim()))
+
+        # If no dynamism is specified, we assume all dimensions are specialized
+        if id(current_inp) not in self.constraints:
+            self._insert_specialized_shapes_assert(arg, all_dims, current_inp)
+            self.input_tracker += 1
+            return arg
+
+        constraints = self.constraints[id(current_inp)]
+
+        constrained_dims = set()
+        # Add runtime asserts for user specified constraints for each
+        # individual dimensions (e.g not the relational constraints like
+        # x[1] == x[0])
```

Comment:
This is not a TOOD right? IIUC even with relational constraints, we will assert on each dim independently?

Reply:
Yep we will. But I think we also should add relational constraints as well. If user has a constrain:
dynamic_dim(x, 0) >= 3, dynamic_dim(x, 1) >= 3, and dynamic_dim(x, 0) == dynamic_dim(x, 1), we still need to error for cases where x.shape[0] = 4 and x.shape[1] = 5.

Reply:
Yes, you're correct. I recall that Avik clarified what equality means in his PR. Ok so it is a TODO to support equality constraints